src/engine/routing_algorithms/alternative_path_mld.cpp

```diff
@@ -0,0 +1,786 @@
+{
+    WeightedViaNode via;
+    UnpackedNodes nodes;
+    UnpackedEdges edges;
+};
+
+// Filters candidates which are on not unique.
+// Returns an iterator to the uniquified range's new end.
+// Note: mutates the range in-place invalidating iterators.
+template <typename RandIt> RandIt filterViaCandidatesByUniqueNodeIds(RandIt first, RandIt last)
+{
+    util::static_assert_iter_category<RandIt, std::random_access_iterator_tag>();
+    util::static_assert_iter_value<RandIt, WeightedViaNode>();
+
+    std::sort(first, last, [](auto lhs, auto rhs) { return lhs.node < rhs.node; });
+    return std::unique(first, last, [](auto lhs, auto rhs) { return lhs.node == rhs.node; });
+}
+
+// Filters candidates which are on un-important roads.
+// Returns an iterator to the filtered range's new end.
+template <typename RandIt>
+RandIt filterViaCandidatesByRoadImportance(RandIt first, RandIt last, const Facade &facade)
+{
+    util::static_assert_iter_category<RandIt, std::random_access_iterator_tag>();
+    util::static_assert_iter_value<RandIt, WeightedViaNode>();
+
+    // Todo: the idea here is to filter out alternatives where the via candidate is not on a
+    // high-priority road. We should experiment if this is really needed or if the boundary
+    // nodes the mld search space gives us already provides us with reasonable via candidates.
+    //
+    // Implementation: we already have `RoadClassification` from guidance. We need to serialize
+    // it to disk and then in the facades read it in again providing `IsImportantRoad(NodeID)`.
+    // Note: serialize out bit vector keyed by node id with 0/1 <=> unimportant/important.
+    (void)first;
+    (void)last;
+    (void)facade;
+
+    return last;
+}
+
+// Filters candidates with much higher weight than the primary route. Mutates range in-place.
+// Returns an iterator to the filtered range's new end.
+template <typename RandIt>
+RandIt filterViaCandidatesByStretch(RandIt first, RandIt last, EdgeWeight weight)
+{
+    util::static_assert_iter_category<RandIt, std::random_access_iterator_tag>();
+    util::static_assert_iter_value<RandIt, WeightedViaNode>();
+
+    // Todo: scale epsilon with weight. Higher epsilon for short routes are reasonable.
+    //
+    //  - shortest path 10 minutes, alternative 13 minutes => 0.30 epsilon Ok
+    //  - shortest path 10 hours, alternative 13 hours     => 0.30 epsilon Unreasonable
+    //
+    // We only have generic weights here and no durations without unpacking.
+    // How do we scale the epsilon in a setup where users can pass us anything as weight?
+
+    const auto stretch_weight_limit = (1. + kEpsilon) * weight;
+
+    const auto over_weight_limit = [=](const auto via) {
+        return via.weight > stretch_weight_limit;
+    };
+
+    return std::remove_if(first, last, over_weight_limit);
+}
+
+// The packed paths' sharing in [0, 1] for no sharing and equality, respectively.
+inline double normalizedPackedPathSharing(const Partition &partition,
```

Comment:
how do you feel about making clear in the function name, that all of these are only heuristics on the packed path?